### PR TITLE
Add a Null2 engine that operates at the core::Engine level

### PR DESCRIPTION
**Pull request by chuckatkins**

### CODE REVIEW

**Comment by germasch** [comment]
This looks nice and encapsulated to me ;)

I have one question, just for my education: Why did you make a separate implementation class? This looks pimpl-like to me, but it's not a typical use case for that (at least not to me). Just being curious -- is it some general separate implementation from interface kind of idea?

**Comment by chuckatkins** [comment]
> Why did you make a separate implementation class? This looks pimpl-like to me

Yep, that's exactly what it is.  


> but it's not a typical use case for that (at least not to me). ... is it some general separate implementation from interface kind of idea?

Yes.  The idea is that it keeps the implementation entirely private even with respect to member variables.

**Comment by williamfgc** [comment]
@chuckatkins sounds good to me, should we call it NullWriter? Thanks.

File: source/adios2/core/IO.cpp [hunk 1]
```diff
@@ -576,6 +577,15 @@ Engine &IO::Open(const std::string &name, const Mode mode,
         engine =
             std::make_shared<engine::NullEngine>(*this, name, mode, mpiComm);
     }
+    else if (engineTypeLC == "null2")
```
Comment (williamfgc) [inline source/adios2/core/IO.cpp]: nullwriter instead of null2?

File: testing/adios2/engine/SmallTestData.h [hunk 2]
```diff
@@ -63,7 +63,7 @@ struct SmallTestData
          std::complex<double>(18.2, 19.2), std::complex<double>(19.2, 20.2)}};
 };
 
-SmallTestData generateNewSmallTestData(SmallTestData in, int step, int rank,
+SmallTestData generateNewSmallTestData(SmallTestData in, size_t step, int rank,
```
Comment (williamfgc) [inline testing/adios2/engine/SmallTestData.h]: I think this is triggering the VS failures. I typically just cast to `int` in the test.

Comment (chuckatkins) [reply testing/adios2/engine/SmallTestData.h]: Yeah, I just didn't have time to look at it last night before I left.  I'll clean it up this morning though.

**Comment by chuckatkins** [comment]
Wrt testing bindings, this is agnostic of the bindings so it shouldn't make any difference.  There's nothing in the bindings that's exercised here that isn't done elsewhere.

File: testing/adios2/engine/SmallTestData.h (hunk 2 above)
Comment (germasch) [inline testing/adios2/engine/SmallTestData.h]: FWIW, this is the change which is causing the CI failure (MSVC, if I remember).

**Comment by germasch** [comment]
FWIW, I think this would be useful for 2.4.0. The CI failure is easy to fix. I'd agree that "NullWriter" is a nicer name than "Null2".

**Comment by chuckatkins** [comment]
I know, it's an easy fix.  i just got way side tracked.  I'll close the loop on it here shortly and get it merged.

File: source/adios2/core/IO.cpp (hunk 1 above)
Comment (chuckatkins) [reply source/adios2/core/IO.cpp]: I've changed it to NullCore

**Pull request merged**